core/src/main/java/org/apache/servicecomb/core/provider/consumer/ConsumerProviderManager.java

```diff
@@ -54,38 +51,32 @@ public void init() throws Exception {
     }
   }
 
-  public ReferenceConfig createReferenceConfig(String microserviceName, String microserviceVersion,
-      String transport) {
-    return new ReferenceConfig(consumerSchemaFactory, microserviceName, microserviceVersion, transport);
+  public ReferenceConfig createReferenceConfig(String microserviceName, String versionRule, String transport) {
+    return new ReferenceConfig(appManager, microserviceName, versionRule, transport);
   }
 
-  public ReferenceConfig getReferenceConfig(String microserviceName) {
-    ReferenceConfig config = referenceConfigMap.get(microserviceName);
-    if (config == null) {
-      synchronized (this) {
-        config = referenceConfigMap.get(microserviceName);
-        if (config == null) {
-          String key = "cse.references." + microserviceName;
-          DynamicStringProperty versionRule = DynamicPropertyFactory.getInstance()
-              .getStringProperty(key + ".version-rule",
-                  DynamicPropertyFactory.getInstance()
-                      .getStringProperty("cse.references.version-rule", Const.VERSION_RULE_LATEST)
-                      .getValue());
-          DynamicStringProperty transport =
-              DynamicPropertyFactory.getInstance().getStringProperty(key + ".transport",
-                  DynamicPropertyFactory.getInstance()
-                      .getStringProperty("cse.references.transport",
-                          Const.ANY_TRANSPORT)
-                      .getValue());
-
-          config = new ReferenceConfig(consumerSchemaFactory, microserviceName, versionRule.getValue(),
-              transport.getValue());
-          referenceConfigMap.put(microserviceName, config);
-        }
-      }
-    }
+  public ReferenceConfig createReferenceConfig(String microserviceName) {
+    String key = "cse.references." + microserviceName;
 
-    return config;
+    String defaultVersionRule = DynamicPropertyFactory.getInstance()
+        .getStringProperty("cse.references.version-rule", DefinitionConst.VERSION_RULE_ALL)
+        .get();
+    String versionRule = DynamicPropertyFactory.getInstance()
+        .getStringProperty(key + ".version-rule", defaultVersionRule)
+        .get();
+
+    String defaultTransport = DynamicPropertyFactory.getInstance()
+        .getStringProperty("cse.references.transport", Const.ANY_TRANSPORT)
+        .get();
+    String transport = DynamicPropertyFactory.getInstance()
+        .getStringProperty(key + ".transport", defaultTransport)
+        .get();
+
+    return new ReferenceConfig(appManager, microserviceName, versionRule, transport);
+  }
+
+  public ReferenceConfig getReferenceConfig(String microserviceName) {
+    return referenceConfigMap.computeIfAbsent(microserviceName, this::createReferenceConfig);
```

Comment:
Not sure if the ConcurrentHashMapEx.computeIfAbsent is atomic,  but if we use the JDK's  one, it's atomic.
https://stackoverflow.com/questions/26481796/concurrenthashmap-computeifabsent

Reply:
ConcurrentHashMapEx extends from ConcurrentHashMap
ConcurrentHashMap.computeIfAbsent even lock when key exists, so there is performance problem

```
  // ConcurrentHashMap.computeIfAbsent always do "synchronized" operation
  // so we wrap it to improve performance
  @Override
  public V computeIfAbsent(K key, Function<? super K, ? extends V> mappingFunction) {
    V value = get(key);
    if (value != null) {
      return value;
    }

    return super.computeIfAbsent(key, mappingFunction);
  }
```